system-tests/src/tests/oauth.spec.ts

```diff
@@ -0,0 +1,439 @@
+import { BrowserContext, expect, Page, test } from "@playwright/test"
+import crypto from "crypto"
+import http from "http"
+import { exportJWK, generateKeyPair, type JWK, SignJWT } from "jose"
+
+type KeyLike = CryptoKey | import("crypto").KeyObject
+// ---------- ENV / CONSTANTS ----------
+const BASE = "http://project-331.local"
+const AUTHORIZE = `${BASE}/api/v0/main-frontend/oauth/authorize`
+const TOKEN = `${BASE}/api/v0/main-frontend/oauth/token`
+const USERINFO = `${BASE}/api/v0/main-frontend/oauth/userinfo`
+const WELL_KNOWN = `${BASE}/api/v0/main-frontend/oauth/.well-known/openid-configuration`
+const JWKS_URI = `${BASE}/api/v0/main-frontend/oauth/jwks.json`
+
+const TEST_CLIENT_ID = "test-client-id"
+const TEST_CLIENT_SECRET = "very-secret" // <- hardcoded as requested
+const APP_DISPLAY_NAME = "Test Client" // shown on consent <h2> and settings <strong>
+const REDIRECT_URI = "http://127.0.0.1:8765/callback" // MUST match the registered redirect for TEST_CLIENT_ID
+
+// Test users
+const USER_EMAIL = "student1@example.com"
+const USER_PASSWORD = "student1"
+const USER_EMAIL_2 = "student2@example.com"
+const USER_PASSWORD_2 = "student2"
+
+// Already-logged-in storage state for the same student
+const STUDENT_STORAGE_STATE = "src/states/student1@example.com.json"
+
+// ---------- CONSENT PAGE OBJECT ----------
```

Comment:
Prefer moving utils to different files (like in utils folder) instead of creating sections in file with comments.